### PR TITLE
What's new in Node 8 and npm 5?

### PR DESCRIPTION
**Pull request by unicodeveloper**

### CODE REVIEW

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 1]
```diff
@@ -0,0 +1,212 @@
+---
+layout: post
+title: "Node 8 and npm@5 release: What's new?"
+description: Node 8.0.0 and npm@5 were released a few days ago. Learn what's new in Node & npm!
+date: 2017-06-03 08:30
+category: Hot Topic, Tools
+design:
+  bg_color: "#1D6A8D"
+  image: https://cdn.auth0.com/blog/yarn-logo.png
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Will you change this logo, right?

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+author:
+  name: Prosper Otemuyiwa
+  url: http://twitter.com/unicodeveloper?lang=en
+  mail: prosper.otemuyiwa@auth0.com
+  avatar: https://en.gravatar.com/avatar/1097492785caf9ffeebffeb624202d8f?s=200
+tags:
+- npm
+- yarn
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Makes sense to add yarn as a tag?

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+- javascript
+- package manager
+related:
+- 2016-09-06-use-nginx-plus-and-auth0-to-authenticate-api-clients
+- 2016-06-29-learn-more-about-our-jwt-chrome-debugger-extension
+---
+
+---
+
+**TL;DR:** Node.js is a JavaScript runtime built on Chrome's V8 JavaScript engine. It has gained massive adoption by developers and organizations around the world because of it's efficient, event-driven and non-blocking I/O model. Node.js makes it possible for developers to code the most popular language (JavaScript) in the world on the backend. Furthermore, Node.js's package manager,`npm`, houses the largest distribution of open source libraries in the world.
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: By reading your "too long, didn't read" section I thought you would talk about what is Node and not what are the new features and improvements on Node.js 8.

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+In this article, we'll highlight 7 notable additions to Node.js and dabble into the new monster, **npm 5**.
+
+## 1. Node.js API (N-API)
+
+Currently, this is an experimental feature behind a flag. This API will eliminate breakage of dependencies that happens between release lines with native modules. 30 percent of JavaScript modules rely indirectly on native modules.
+
+> Native modules are modules written in C or C++ and directly bound to the Chrome V8.
+
+Native modules builders and users can start experimenting with this feature. More details can be found [here](https://medium.com/@nodejs/n-api-next-generation-node-js-apis-for-native-modules-169af5235b06).
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: I still don't get what N-API is actually. Perhaps you could explain a little bit more.

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+## 2. Better Support for Promises
+
+This release includes a new `util.promisify()` API that allows developers to wrap standard callback-style APIs in a function that returns a Promise. Check out this sample code:
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Add a link to some docs.

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+readfile('/some/file')
+  .then((data) => { /** ... **/ })
+  .catch((err) => { /** ... **/ });
+```
+
+## 3. Buffer Improvements
+
+The Node.js Buffer API witnessed a lot of changes. Most notable amongst these changes is:
+
+* Zero-filling new instances of Buffer(num) by default will have a significant impact on performance.
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Perhaps you should say what kind of impact. By reading this, one may think that these changes will bring a good impact, which is not true. Furthermore, if you say that it will loose some performance, it is important to say why are they doing this.

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+// Uninitialized Buffer
+const unsafeBuffer = Buffer.allocUnsafe(20);
+```
+
+> **Note:** Developers should be very careful when using the `Buffer.allocUnsafe`. Only use it when you are aware of the risks and how to avoid it.
+
+## 4. Stable WHATWG URL Parser
+
+This release makes the **WHATWG URL** parser fully supported. No more hiding behind the experimental flag. It's a URL API implementation that matches the `URL` implementation in modern web browsers like Firefox, Edge, Chrome, and Safari allowing code using URLs to be shared across environments.
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: What do you mean by "allowing code using URLs to be shared across environments"? You mean that some code using `URL` (like the example below) will be handled equally on these different browsers?

Just asking because I'm unsure if I understood correctly.

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+```js
+const async_hooks = require('async_hooks');
+
+const cid = async_hooks.currentId();
+// returns the ID of the current execution context.
+
+const tid = async_hooks.triggerId();
+// returns the ID of the handle responsible for triggering the callback of the current execution scope to call.
+```
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Could you give a more enlightening example? I read it and I still don't know why does this matter and what I could do with this...

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+The full documentation can be found [here](https://github.com/nodejs/node/pull/13287).
+
+## 6. Stream API Improvements
+
+This release adds new ways for destroying and finalizing Stream instances. Developers can now write custom implementations for the `destroy()` method that every Stream instance inherits.
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Perhaps some example and also an explanation on why do you consider this a "notable addition"? I'm not sure why would I care about creating my own `destroy()` method for streams.

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+  // Do whatever
+});
+
+// Send messages to the inspector
+session.post(message);
+
+session.disconnect();
+```
+
+Check out the full list of [changes here](https://nodejs.org/en/blog/release/v8.0.0/).
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Not a very useful link. Adding "#experimental-inspector-javascript-api" would be more precise, but yet I didn't see any "full list" there. Oh, now I see. You mean the full list of changes on Node.js 8, not for the "Inspector JavaScript API"...

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+Developers have a new way of debugging their Node.js applications via the experimental inspector JavaScript API. This API leverages the debug protocol to inspect various Node.js processes.
+
+```js
+const inspector = require('inspector');
+
+const session = new inspector.Session();
+session.connect();
+
+// Listen for inspector events
+session.on('inspectorNotification', (message) => {
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Are there other events? What are the applicabilities of this inspector API? Why would anyone use it and how?

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+
+### 2. Lockfiles
+
+With `npm@5`, lockfiles are the default (`package-lock.json`). This simply means that whatever files you get when you install a package will be the same every time you install that package after initial install. This eliminates the challenges developers had with having different files on different developer environments after installing the same package.
+
+![lockfiles](https://cdn.auth0.com/blog/npm5/package-lock.png)
+
+### 3. SHA-512 Hash
+
+The npm CLI will use `sha512` for tarballs from registries that sends a `sha512` checksum as the tarball hash. `npm@5` publishes with SHA-512 hashes. This is to protect a developer against malicious attacks by checking downloaded packages and ensuring that code downloaded from the registry is safe and unwavering!
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: So, before npm 5, no hash validations were made?

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+
+## npm@5
+
+**npm's** latest version is `npm@5`. What's new in **npm@5**?
+
+### 1. Wicked Fast Install
+
+I bumped my npm version to npm 5 using `npm install npm@latest -g` and the first thing I discovered was that my module installs were insanely fast!
+
+![Speed improvements](https://cdn-images-1.medium.com/max/1600/0*K1Wb1ERhtAHLRG0m.)
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: How did they achieve these improvements? Besides that, is NPM now faster than Yarn? The opposite? Or do they have similar performance?

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+`npm@5` doesn't show the whole installed tree on your screen anymore. Instead, you'll see a summary report of the install on your screen like so:
+
+![Install report summary](https://cdn.auth0.com/blog/npm5/summary.png)
+
+`Added 2 packages in 0.599s`
+
+### 6. Automatic Fallback to Offline mode
+
+`npm@5` will use your cache if you are offline to install packages that have been installed before but are no longer available. Check this out:
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: I don't get it, what do you mean by "but are no longer available"? Is it related to the fact that the developer is offline? Or are you trying to say that this fallback will be useful in the following two situations:
1. When a dev has no internet connection.
2. When the package has (for some reason) been removed from NPM.

?

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+```bash
+npm install auth0-lock
+```
+
+It installs the widget within seconds and locks down the exact files with `package-lock.json` file. If you don't already have an Auth0 account, [sign up](javascript:signup\(\)) for one now. Navigate to the Auth0 [management dashboard](https://manage.auth0.com/), select **Applications** from the navigational menu, then select the app you want to connect with the JavaScript framework of your choice. Now head over to the [Quickstart docs](https://auth0.com/docs/quickstarts), select the type of app you want to build and follow the steps highlighted there.
+
+
+## Conclusion
+
+**Node.js 8** and **npm5** came loaded with lots of new features and significant improvements. I couldn't be more proud of the JavaScript and Open source community right now. Is [Yarn](https://github.com/yarnpkg/yarn)going to release something super amazing soon? Maybe a faster install experience? I'll be on the look out!
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: You are missing a space between the word "Yarn" and the word "going".

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 1 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Yes

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 2 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Yes

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 3 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Is this not good enough description of what Node is?

>  Node.js is a JavaScript runtime built on Chrome's V8 JavaScript engine.

Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Also

>  Node.js makes it possible for developers to code the most popular language (JavaScript) in the world on the backend.

@brunokrebs

Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: What I mean is that TL;DR, as far as I know, is meant to provide a summary about the article that is coming. Which is not the case here. But anyway, as you prefer.

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 5 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: True

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 6 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Make sense

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 7 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Yes, Exactly what I mean. I will probably use your English here for better clarification

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 8 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: The documentation for this method is not available right now. I had to pull this from the source code itself. But does the explanation I gave make sense?

It's a hook that you add to your code to get information about whatever is happening within the scope of any function or callback you specify the hook in?

@brunokrebs

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 10 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Yea exactly

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 13 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Yes, NPM is now faster than Yarn

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 12 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: I'm really not sure about this

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 14 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: 1. When the dev has no internet connection.
2. If the node_modules folder gets deleted and the package is no longer available.

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 15 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Oh thanks

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 8 above)
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Well, yes, it does make sense. But it is a little bit hard to see what one can really do with this. That's why I asked if a more complete/enlightening example would be possible.

I don't know what `async_hooks` really do and, by reading this article, I think that it is a "global" way of defining listeners for asynchronous resources. But I'm still unsure.

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 3 above)
Comment (diegopoza) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: I agree with @brunokrebs, it is nice to have as small description of what node is, but I would also try to provide a little more info on the article that is coming. @unicodeveloper

Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Alright. I will add that @diegopoza

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 16]
```diff
@@ -25,25 +25,27 @@ related:
 
 ---
 
-**TL;DR:** Node.js is a JavaScript runtime built on Chrome's V8 JavaScript engine. It has gained massive adoption by developers and organizations around the world because of it's efficient, event-driven and non-blocking I/O model. Node.js makes it possible for developers to code the most popular language (JavaScript) in the world on the backend. Furthermore, Node.js's package manager,`npm`, houses the largest distribution of open source libraries in the world.
+**TL;DR:** Node.js is a JavaScript runtime built on Chrome's V8 JavaScript engine. It has gained massive adoption by developers and organizations around the world because of it's efficient, event-driven and non-blocking I/O model. Node.js makes it possible for developers to code the most popular language (JavaScript) in the world on the backend. Furthermore, Node.js's package manager,`npm`, houses the largest distribution of open source libraries in the world. Node.js and npm just had new major releases. In this article, we'll highlight 7 notable additions to Node.js and dabble into the new monst, **npm@5**.
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: What does monst means? Is it supposed to be monster?

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown [hunk 17]
```diff
@@ -25,25 +25,27 @@ related:
...
+Node.js 8.0.0 was announced to the world on May 31, 2017. The previous Node.js version was v7.10.0. The *v* was dropped in order to avoid confusion with the V8 JavaScript engine. This major release brings a lot of new features, deprecations and changes. By October, 2017, Node.js 8 will become a Long Term Support (LTS) release. Next, notable additions!
 
 ## 1. Node.js API (N-API)
 
-Currently, this is an experimental feature behind a flag. This API will eliminate breakage of dependencies that happens between release lines with native modules. 30 percent of JavaScript modules rely indirectly on native modules.
+Currently, this is an experimental feature behind a flag. 30 percent of JavaScript modules rely indirectly on native modules.
+
+Existing native modules are written in C/C++ and directly depend on V8 or [NAN](https://github.com/nodejs/nan) APIs. This dependency requires native modules to be recompiled and updated for every major Node.js release. This is a burden for native modules maintainers. Furthermore, companies and products depending on these modules find it difficult to upgrade their Node.js versions in production because of potential massive code breakage.
```
Comment (brunokrebs) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: I would replace "NAN" for "Native Abstractions for Node.js (NAN)"

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 16 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Yeah, monster. My bad!

File: _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown (hunk 17 above)
Comment (unicodeveloper) [inline _posts/2017-06-02-whats-new-in-node8-and-npm5.markdown]: Alright
